service/worker/scheduler/workflow_test.go

```diff
@@ -1311,6 +1320,140 @@ func (s *workflowSuite) TestBackfillInclusiveStartEnd() {
 	)
 }
 
+func (s *workflowSuite) TestHugeBackfillAllowAll() {
+	// TODO: remove once default version is IncrementalBackfill
+	currentVersion := currentTweakablePolicies.Version
+	currentTweakablePolicies.Version = IncrementalBackfill
+	defer func() { currentTweakablePolicies.Version = currentVersion }()
+
+	// This has been run for up to 5000, but it takes a very long time. Run only 100 normally.
+	const backfillRuns = 100
+	const backfills = 4
+	// The number that we process per iteration, with a 1s sleep per iteration, makes an
+	// effective "rate limit" for processing allow-all backfills. This is different from the
+	// explicit rate limit.
+	rateLimit := currentTweakablePolicies.BackfillsPerIteration
+
+	base := time.Date(2001, 8, 6, 0, 0, 0, 0, time.UTC)
+	runs := make([]workflowRun, backfillRuns)
+	for i := range runs {
+		t := base.Add(time.Duration(i) * time.Hour)
+		actual := baseStartTime.Add(time.Minute).Add(time.Duration(i/rateLimit) * time.Second)
+		runs[i] = workflowRun{
+			id:             "myid-" + t.Format(time.RFC3339),
+			start:          actual,
+			startTolerance: 5 * time.Second, // the "rate limit" isn't exact
+			end:            actual.Add(time.Minute),
+			result:         enumspb.WORKFLOW_EXECUTION_STATUS_COMPLETED,
+		}
+	}
+
+	delayedCallbacks := make([]delayedCallback, backfills)
+	for i := range delayedCallbacks {
+		i := i
+		delayedCallbacks[i] = delayedCallback{
+			// test environment seems to get confused if the callback falls on the same instant
+			// as a workflow timer, so use an odd interval to force it to be different.
+			at: baseStartTime.Add(time.Minute).Add(time.Duration(i) * 1113 * time.Millisecond),
+			f: func() {
+				s.env.SignalWorkflow(SignalNamePatch, &schedpb.SchedulePatch{
+					BackfillRequest: []*schedpb.BackfillRequest{{
+						StartTime:     timestamppb.New(base.Add(time.Duration(i*backfillRuns/backfills) * time.Hour)),
+						EndTime:       timestamppb.New(base.Add(time.Duration((i+1)*backfillRuns/backfills-1) * time.Hour)),
+						OverlapPolicy: enumspb.SCHEDULE_OVERLAP_POLICY_ALLOW_ALL,
+					}},
+				})
+			},
+		}
+	}
+
+	delayedCallbacks = append(delayedCallbacks, delayedCallback{
+		at:         baseStartTime.Add(50 * time.Minute),
+		finishTest: true,
+	})
+
+	s.runAcrossContinue(
+		runs,
+		delayedCallbacks,
+		&schedpb.Schedule{
+			Spec: &schedpb.ScheduleSpec{
+				Interval: []*schedpb.IntervalSpec{{Interval: durationpb.New(time.Hour)}},
+			},
+		},
+	)
+}
+
+func (s *workflowSuite) TestHugeBackfillBuffer() {
+	// TODO: remove once default version is IncrementalBackfill
+	prevTweakables := currentTweakablePolicies
+	currentTweakablePolicies.Version = IncrementalBackfill
+	currentTweakablePolicies.MaxBufferSize = 30 // make smaller for testing
+	defer func() { currentTweakablePolicies = prevTweakables }()
+
+	// This has been run for up to 3000, but it takes a very long time. Run only 100 normally.
+	const backfillRuns = 100
+	const backfills = 4
```

Comment:
Could we inject or configure the sleep duration when testing so that we don't have to care?

Reply:
It's already running in the workflow test framework which does time skipping. The bookkeeping in the test framework is just really really slow (and there are some quadratic bits in there.. elapsed time scales with the square of backfullRuns). Probably at least some of the quadraticness is from the mocks.